src/fges.jl

```diff
@@ -0,0 +1,536 @@
+            
+            if newScore > bestScore
+                bestH = H
+                bestScore = newScore
+            end
+        end
+    end
+
+    return (bestH, bestScore)
+end
+
+
+####################################################################
+# Scoring function
+####################################################################
+
+@memoize LRU(maxsize=1_000_000) function score(dataParsed::ParseData{Matrix{A}}, nodeParents, node) where A
+
+    # Unpack some variables from the dataParsed structure
+    n = A(dataParsed.numObservations) #convert datatype
+    data = dataParsed.normAugData
+    p = dataParsed.penalty
+
+    # The last column of dataParsed.normAugData is all ones which is required for a linear regression with an intercept. If there are no node parents, model the child node with just the intercept, else use the parents and the intercept
+    if isempty(nodeParents)
+        parentsAndIncept = [dataParsed.numFeatures+1]
+    else
+        parentsAndIncept = [nodeParents; dataParsed.numFeatures+1]
+    end
+
+    # To calculate the score we need a mean-squared error which we can get by regessing the the child node onto the parents
+    # Create variables for a linear regression y = X*b
+
+    # Use views to avoid creating copies of the data in memory
+        # X is the design matrix, augmented with a column of ones at the end
+        # X is also been standardized so mean(columns)=0 and std(column)=1
+        # y is data from the child node being tested
+    @views begin
+        y = data[:,node]
+        X = data[:,parentsAndIncept]
+    end
+
+    # Perform a linear regression
+    b = X \ y
+
+    # Get the estimation
+    ŷ = X*b
+
+    # Next we want to calculate the log likelihood that these are the parents of our node
+    # score = log(P(data|Model)) ≈ -BIC/2
+    # because we're only comparing log likelihoods we'll ignore the 1/2 factor
+    # when P(⋅) is Guassian, log(P(data|Model)) takes the form:
+    # -k⋅log(n) - n⋅log(mse)
+    # k is the number of free parameters and mse is mean squared error
+    k = length(parentsAndIncept) #includes the intercept
+    mse = sum(x->x^2, y-ŷ) / n
+
+    # return the final score we want to maximize (which is technically -2BIC)
+    return -p*k*log(n) - n*log(mse)
+end
+
+
+
+####################################################################
+# Using Meek's Rules to Update PDAG
+####################################################################
+
+# Revert a graph to undirected edges and unshielded colliders (i.e. parents not adjacent)
+function graphVStructure!(g)
+    
+    # loop through all vertices
+    for x in vertices(g)
+        # get the parents of current vertex
+        parentsX = parents(g,x)
+        # if all the parents are adjacent, undirect the edges
+        # (if there is only 1 parent, then it is still a clique)
+        if isclique(g, parentsX)
```

Comment:
I'm not sure this works as intendend. E.g. if you have x with three parents p1, p2 and p3 which are connected p1 - p2 - p3. Then, they do not form a clique, but p2 is not part of a v-structure so p2 - x should be made undirected?

Reply:
This function can be tested with https://github.com/mschauer/CausalInference.jl/blob/82964f4c716e92ee37a45fd5f446fcba00b436a4/src/pc.jl#L111 which does the same using the indirect method of running the oracle pc algorithm.

Reply:
@mwien you are correct. Another way may be to simply loop over all directed edges and check if they don't participate in an unshielded collider.

```julia
function graphVStructure!(g)
    
    #undirect an edge if it does not participate in an unshielded collider
    for edge in edges(g)
        if isdirected(g,edge) && allshielded(g,edge) 
            #undirect by adding reverse edge
            add_edge!(g, edge.child, edge.parent)
        end
    end
end

isdirected(g, edge) = has_edge(g,edge) ⊻ has_edge(g,reverse(edge)) # xor

#x → y ← p check that x is adjacent to all p
allshielded(g,x,y) = all(isadjacent(g, p, x) for p in parents(g, y) if p ≠ x)
allshielded(g,edge) = allshielded(g, edge.src, edge.dst)
```

Reply:
Newly implemented